src/iterators/mod.rs

```diff
@@ -155,50 +170,53 @@ impl<'a, T: 'a, P> Iterator for HistogramIterator<'a, T, P>
                 return None;
             }
 
-            // TODO should check if we've reached max, not count, to avoid early termination
-            // on histograms with very large counts whose total would exceed u64::max_value()
-            // Have we yielded all non-zeros in the histogram?
-            let total = self.hist.count();
-            if self.prev_total_count == total {
+            // Have we already picked the index with the last non-zero count in the histogram?
+            if self.last_picked_index >= self.max_value_index {
                 // is the picker done?
                 if !self.picker.more(self.current_index) {
                     self.ended = true;
                     return None;
                 }
-
-            // nope -- alright, let's keep iterating
             } else {
+                // nope -- alright, let's keep iterating
                 assert!(self.current_index < self.hist.len());
-                assert!(self.prev_total_count < total);
 
                 if self.fresh {
-                    let count = self.hist.count_at_index(self.current_index)
-                        .expect("Already checked that current_index is < counts len");
-
-                    // if we've seen all counts, no other counts should be non-zero
-                    if self.total_count_to_index == total {
-                        // TODO this can fail when total count overflows
-                        assert_eq!(count, T::zero());
-                    }
+                    // at a new index, and not past the max, so there's nonzero counts to add
+                    self.count_at_index = self.hist.count_at_index(self.current_index)
+                            .expect("Already checked that current_index is < counts len");
 
-                    // TODO overflow
-                    self.total_count_to_index = self.total_count_to_index + count.as_u64();
+                    self.total_count_to_index =
+                            self.total_count_to_index.saturating_add(self.count_at_index.as_u64());
+                    self.count_since_last_iteration =
+                            self.count_since_last_iteration.saturating_add(self.count_at_index.as_u64());
 
                     // make sure we don't add this index again
                     self.fresh = false;
                 }
             }
 
             // figure out if picker thinks we should yield this value
-            if self.picker.pick(self.current_index, self.total_count_to_index) {
-                let val = self.current();
```

Comment:
I felt it was easier to reason at-a-glance about when the iterator's fields were used with this function inlined, since it matters that you read things like `count_since_last_iteration` *before* resetting a few lines below.